batcher/message_batcher.go

```diff
@@ -0,0 +1,145 @@
+package batcher
+
+import (
+	"fmt"
+	"math/big"
+	"time"
+)
+
+type Sync interface {
+	SendBatch(batch [][]byte, largestSeq *big.Int, largetsSubSeq int)
+}
+
+type Batcher interface {
+	// Interval at which accumulated messages should be bulk put to
+	// firehose (default 1 second).
+	FlushInterval(dur time.Duration)
+	// Number of messages that triggers a push to firehose
+	// default to 10
+	FlushCount(count int)
+	// Size of batch that triggers a push to firehose
+	// default to 1mb (1024 * 1024)
+	FlushSize(size int)
+
+	//AddMesage to the batch
+	AddMessage(msg []byte, sequenceNumber string, subSequenceNumber int) error
+
+	//Flush all messages from the batch
+	Flush()
+}
+
+type batcher struct {
+	flushInterval time.Duration
+	flushCount    int
+	flushSize     int
+
+	largestSeq    *big.Int
+	largestSubSeq int
+
+	sync      Sync
+	msgChan   chan<- []byte
+	flushChan chan<- struct{}
+}
+
+func New(sync Sync) *batcher {
+	msgChan := make(chan []byte, 100)
+	flushChan := make(chan struct{})
+
+	b := &batcher{
+		flushCount:    10,
+		flushInterval: time.Second,
+		flushSize:     1024 * 1024,
+
+		sync:      sync,
+		msgChan:   msgChan,
+		flushChan: flushChan,
+	}
+
+	go b.startBatcher(msgChan, flushChan)
+
+	return b
+}
+
+func (b *batcher) FlushInterval(dur time.Duration) {
+	b.flushInterval = dur
+}
+
+func (b *batcher) FlushCount(count int) {
+	b.flushCount = count
+}
+
+func (b *batcher) FlushSize(size int) {
+	b.flushSize = size
+}
+
+func (b *batcher) AddMessage(msg []byte, sequenceNumber string, subSequenceNumber int) error {
+	if len(msg) <= 0 {
+		return fmt.Errorf("Empty messages can't be sent")
+	}
+
+	b.msgChan <- msg
+	err := b.updateSequenceNumber(sequenceNumber, subSequenceNumber)
+	if err != nil {
+		return err
+	}
+	return nil
+}
+
+// updateSequenceNumber updates sequence number on the batch, if a larger one is available
```

Comment:
"updates sequence number on the batch" could be improved... it's not clear to me what it means